File: 5-Longest Palindromic Substring.cpp
class Solution {
public:
	string longestPalindrome(string s) {
		string ss = insert(s);
		int dp[2001];
		dp[0] = 1;
		dp[1] = 2;

		int maxRight = 2, maxLength = 2;
		int len = ss.size();
		for (int i = 2; i < len; ++i) {
			int left = 2 * (maxRight - maxLength + 1) - i;
			int begin;
			if (i >= maxRight) {
				begin = i + 1;
			}
			else {
				begin = dp[left] - 1 < maxRight - i ? i + dp[left] : maxRight + 1;
			}

			while (begin < len && 2 * i - begin >= 0
				&& ss[begin] == ss[2 * i - begin])
			{
				++begin;
			}

			dp[i] = begin - i;
			if (dp[i] > maxLength) {
				maxLength = dp[i];
				maxRight = i + dp[i] - 1;
			}
		}

		int begin = maxRight - 2 * (maxLength - 1);
		string result = "";
		for (int i = begin + 1; i <= maxRight; i += 2) {
			result += ss[i];
		}
		return result;
	}
private:
	string insert(const string &s) {
		int len = s.size();
		string result = "#";
		for (int i = 0; i < len; ++i) {
			result += s[i];
			result += "#";
		}
		return result;
	}
};

File: 5-Longest Palindromic Substring.java
class Solution {
    public String longestPalindrome(String s) {
        String str = preprocess(s);
        
        int len = str.length();
        int[] dp = new int[len];
        int center = 0, boundary = 0, left = 0, right = 0;
        for (int i = 1; i < len; ++i) {
            dp[i] = i >= boundary ? 0 : Math.min(boundary - i, dp[2 * center - i]);
            
            while (i - 1 - dp[i] >= 0 && i + 1 + dp[i] < len && str.charAt(i - 1 - dp[i]) == str.charAt(i + 1 + dp[i]))
                ++dp[i];
            
            if (right - left < dp[i] * 2 + 1) { right = i + dp[i]; left = i - dp[i]; }
            if (boundary < i + dp[i]) { boundary = i + dp[i]; center = i; }
        }
        
        return s.substring(left >> 1, right >> 1);
    }
    
    private String preprocess(String s) {
        StringBuilder builder = new StringBuilder(s.length() * 2 + 1);
        for (char c : s.toCharArray()) {
            builder.append('#').append(c);
        }
        builder.append('#');
        return builder.toString();
    }
}
